Mathlib/Tactic/NormCast.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2019 Paul-Nicolas Madelaine. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul-Nicolas Madelaine, Robert Y. Lewis, Mario Carneiro
+-/
+
+import Lean
+import Mathlib.Tactic.RunCmd
+
+/-!
+# Stub for implementation of the `@[norm_cast]` attribute.
+
+This file is currently just a stub that creates a no-operation `@[norm_cast]` attribute.
+Without this, all declarations in the mathport output for mathlib3 that use `@[norm_cast]` fail.
+With the no-operation attribute, the declarations can succeed,
+but of course all later proofs that rely on the existence of the automatically generated lemmas
+will fail.
+
+Later we will need to port the implementation from mathlib3.
+
+-/
+
+
+open Lean Meta
+
+namespace Lean.Parser.Attr
+
+syntax (name := normCast) "norm_cast" (ppSpace (&"elim" <|> &"move" <|> &"squash"))? : attr
+
+end Lean.Parser.Attr
+
+namespace Tactic.NormCast
+
+initialize pushCastExtension : SimpExtension ← registerSimpAttr `push_cast $
+  "The `push_cast` simp attribute uses `norm_cast` lemmas " ++
+  "to move casts toward the leaf nodes of the expression."
+
+/--  The cache for `norm_cast` attribute stores three `SimpLemmas` objects. -/
+structure NormCastExtension where
+  up : SimpExtension
+  down : SimpExtension
+  squash : SimpExtension
+  deriving Inhabited
+
+builtin_initialize normCastExt : NormCastExtension ← pure {
+  up := ← mkSimpExt `Tactic.NormCast.normCastExt.up
```

Comment:
I would normally use  ``` ``normCastExt``` here, but it's the definition being declared so it's not in scope yet. I believe these names are normally fully namespaced.

`push_cast` is different, it's calling `mkSimpAttr` instead of `mkSimpExt`. The name it accepts is the name of the attribute, so it is user facing (i.e. people write `@[push_cast]`). `mkSimpAttr` creates a derived name for the extension, although it doesn't appear to be namespaced. With `mkSimpExt` alone, you can't put things in the simp set by using an attribute at all, which is what we want since they are internal and `norm_cast` has its own attribute.